filesetup.c

```diff
@@ -338,6 +338,99 @@ static bool pre_read_file(struct thread_data *td, struct fio_file *f)
 	return ret;
 }
 
+/*
+ * Generic function to prepopulate regular file with data.
+ * Useful if you want to make sure I/O engine has data to read.
+ * Leaves f->fd open on success, caller must close.
+ */
+int generic_prepopulate_file(struct thread_data *td, struct fio_file *f)
+{
+	int flags;
+	unsigned long long left, bs;
+	char *b = NULL;
+
+	/* generic function for regular files only */
+	assert(f->filetype == FIO_TYPE_FILE);
+
+	if (read_only) {
+		log_err("fio: refusing to write a file due to read-only\n");
+		return 0;
+	}
+	if (td->o.fill_device) {
+		log_err("fio: prepopulatig device when fill_device is set, makes no sense\n");
+		return 0;
+	}
+
+	flags = O_WRONLY;
+	if (td->o.allow_create)
+		flags |= O_CREAT;
+
+#ifdef WIN32
+	flags |= _O_BINARY;
+#endif
+
+	dprint(FD_FILE, "open file %s, flags %x\n", f->file_name, flags);
+	f->fd = open(f->file_name, flags, 0644);
+	if (f->fd < 0) {
+		int err = errno;
+
+		if (err == ENOENT && !td->o.allow_create)
+			log_err("fio: file creation disallowed by "
+					"allow_file_create=0\n");
+		else
+			td_verror(td, err, "open");
+		return 1;
+	}
+
+	left = f->real_file_size; // XXX: ?? f->file_offset + f->io_size
```

Comment:
I'm not sure - if I use the offset + io_size, how to apply offset into the write...?

Reply:
f->real_file_size should be the actual size of the file, f->file_offset is just where IO would be done. So I think you should lay out as f->real_file_size as you do.

Reply:
Ok, I've removed the comment.